modules/gapi/src/backends/onnx/gonnxbackend.cpp

```diff
@@ -242,12 +255,35 @@ inline void preprocess(const cv::Mat& src,
             dst = pp;
         }
 
+        if (with_pad) {
+            const auto div = ti.pad->divisibility;
+            const int padded_h = std::ceil(new_h / div) * div;
+            const int padded_w = std::ceil(new_w / div) * div;
+            int type = -1;
+            // is_hwc - Mat has C*H, W dimensions
+            if (!is_hwc) {
+                type = ddepth;
+            } else {
+                type = CV_MAKETYPE(ddepth, new_c);
+            }
+            GAPI_Assert(type != -1);
+            cv::Mat pad_im(cv::Size(padded_w, (!is_hwc ? new_c : 1) * padded_h), type, 0.f);
+            pad_im(cv::Rect(0, 0, dst.cols, dst.rows)) += dst;
+            dst = pad_im;
+            new_h = padded_h;
+            new_w = padded_w;
+        }
+
         // Ensure dst is a tensor shape (not a 2D image)
         if (ti.is_dynamic) {
             // Reshape to input dimensions
             const std::vector<int> out_dims = is_hwc
-                ? std::vector<int>{1, new_h, new_w, new_c}
-                : std::vector<int>{1, new_c, new_h, new_w};
+                ? with_batch
+                    ? std::vector<int>{1, new_h, new_w, new_c}
+                    : std::vector<int>{new_h, new_w, new_c}
+                : with_batch
+                    ? std::vector<int>{1, new_c, new_h, new_w}
+                    : std::vector<int>{new_c, new_h, new_w};
```

Comment:
Does ONNX expect a 3-dim vector in that particular case?

Reply:
`with_batch = ti.dims.size() == 4u ? true : false;`
where `ti` - tensor info, which contains `dims` obtained from network parameters (`session.GetInputTypeInfo(...).GetTensorTypeAndShapeInfo().GetShape()`).
Mask/FasterRCNN expects a 3-dim input.